File: c/2015/day05b/day05b.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void usage(char *progname) {
    fprintf(stderr, "usage: %s <input file>\n", progname);
    exit(1);
}

int prop1(char *str) {
    char first[3], second[3];
    int i, j;

    for ( i = 0; i < strlen(str) - 4; i++ ) {
        strncpy(first, str + i, 2);
	for ( j = i + 2; j < strlen(str) - 2; j++ ) {
	    strncpy(second, str + j, 2);
	    if ( strncmp(first, second, 2) == 0 ) {
		return 1;
	    }
	}
    }
    return 0;
}

int prop2(char *str) {
    char *p;

    for ( p = str + 2; *p != '\0'; p++ ) {
        if ( *p == *(p - 2) ) {
            return 1;
        }
    }
    return 0;
}

int process(char *filename) {
    FILE *infile;
    char *line;
    size_t size;
    int count = 0;

    if ( (infile = fopen(filename, "r")) == NULL ) {
        perror("failed to open input file");
        exit(1);
    }
    line = NULL;
    while ( getline(&line, &size, infile) != -1 ) {
        if ( prop1(line) && prop2(line) ) {
                count++;
        }
        free(line);
        line = NULL;
    }
    fclose(infile);
    return count;
}

int main(int argc, char *argv[]) {
    char *filename;
    int result;

    if ( argc < 2 ) {
        usage(argv[0]);
    }

    filename = argv[1];
    result = process(filename);
    printf("result = %d\n", result);
    return 0;
}

File: pascal/2015/day05b/day05b.pas
PROGRAM Day05a(Input, Output);

USES
    Math;

VAR
    FileName : PChar;
    Res      : UInt32;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Prop1(S : String) : Boolean;
VAR
    I, J                 : SizeInt;
    Slice1, Slice2       : String;
    Found                : Boolean;

BEGIN
    Found := False;
    I := 1;
    WHILE (I < Length(S) - 2) AND NOT Found DO
    BEGIN
        Slice1 := Copy(S, I, 2);
        J := I + 2;
        WHILE (J < Length(S)) AND NOT Found DO
        BEGIN
            Slice2 := Copy(S, J, 2);
            IF Slice1 = Slice2 THEN
                Found := True
            ELSE
                J := J + 1
        END;
        IF NOT Found THEN
            I := I + 1;
    END;
    Prop1 := Found
END;

FUNCTION Prop2(S : String) : Boolean;
VAR
    I                    : SizeInt;
    Found                : Boolean;

BEGIN
    Found := False;
    I := 1;
    WHILE (I < Length(S) - 1) AND NOT Found DO
        IF S[I] = S[I + 2] THEN
            Found := True
        ELSE
            I := I + 1;
    Prop2 := Found
END;

FUNCTION Process(FileName : PChar) : UInt32;
VAR
    InFile               : Text;
    LineStr              : String;
    Count                : UInt32;

BEGIN
    Count := 0;
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        ReadLn(InFile, LineStr);
        IF Prop1(LineStr) AND Prop2(LineStr) THEN
            Count := Count + 1
    END;
    Close(InFile);
    Process := Count
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.
